Revision:34689

Bacchusisi opened the pull request:
【bug】【java驱动】【游标泄露】检查java驱动存在游标泄露的情况
【问题描述】 
Java驱动中使用游标存在没有主动关闭的情况。 

【存在问题的地方】 
1、检查驱动所有代码，只发现Sequoiadb#isDomainExist明确存在游标泄露的情况。 
2、其它地方存在没有主动关闭游标的情况，但正常情况下不会发生游标泄露。 

【问题修改】 
将所有使用到游标的地方都进行主动游标关闭。

Bacchusisi commented:
![image](https://user-images.githubusercontent.com/47816465/56185588-0b9d0200-6050-11e9-8a30-38b73533ef4d.png)

Bacchusisi commented:
【提交】 
```
v2.8 
Command: Commit 
Modified: F:\sequoiadb_2.8\driver\java\src\main\java\com\sequoiadb\base\CollectionSpace.java 
Modified: F:\sequoiadb_2.8\driver\java\src\main\java\com\sequoiadb\base\DBCollection.java 
Modified: F:\sequoiadb_2.8\driver\java\src\main\java\com\sequoiadb\base\DBDataCenter.java 
Modified: F:\sequoiadb_2.8\driver\java\src\main\java\com\sequoiadb\base\Node.java 
Modified: F:\sequoiadb_2.8\driver\java\src\main\java\com\sequoiadb\base\ReplicaGroup.java 
Modified: F:\sequoiadb_2.8\driver\java\src\main\java\com\sequoiadb\base\Sequoiadb.java 
Sending content: F:\sequoiadb_2.8\driver\java\src\main\java\com\sequoiadb\base\DBDataCenter.java 
Sending content: F:\sequoiadb_2.8\driver\java\src\main\java\com\sequoiadb\base\DBCollection.java 
Sending content: F:\sequoiadb_2.8\driver\java\src\main\java\com\sequoiadb\base\Node.java 
Sending content: F:\sequoiadb_2.8\driver\java\src\main\java\com\sequoiadb\base\CollectionSpace.java 
Sending content: F:\sequoiadb_2.8\driver\java\src\main\java\com\sequoiadb\base\ReplicaGroup.java 
Sending content: F:\sequoiadb_2.8\driver\java\src\main\java\com\sequoiadb\base\Sequoiadb.java 
```
【check list】 
检查项结论（NA/N/Y）
修改代码是否全部提交SVN主干和分支版本且符合代码规范--- v2.8 
是否用例全部通过-------------------------------------- Y 
是否清淅标明重现步骤---------------------------------- NA 
修改代码是否通过检视---------------------------------- N 
资料是否变更------------------------------------------ NA 
接口是否变更------------------------------------------ N 
相关驱动是否变更或添加-------------------------------- NA 
是否保持向前兼容-------------------------------------- Y 
是否影响安装部署和监控-------------------------------- NA 
是否对影响性能---------------------------------------- NA 
sdb shell帮助信息是否更新----------------------------- NA

Bacchusisi commented:
请安排检视和测试。

FlorentinoJH commented:
请检视

Bacchusisi commented:
【提交】 
```
v2.9 
Command: Commit 
Modified: E:\sequoiadb\driver\java\src\main\java\com\sequoiadb\base\DBCollection.java 
Modified: E:\sequoiadb\driver\java\src\main\java\com\sequoiadb\base\Sequoiadb.java 
Modified: E:\sequoiadb\driver\java\src\test\java\com\sequoiadb\test\TestConfig.java 
Modified: E:\sequoiadb\driver\java\src\test\java\com\sequoiadb\test\bug\Bug_JIRA_.java 
Modified: E:\sequoiadb\driver\java\src\test\java\com\sequoiadb\test\common\Constants.java 
Modified: E:\sequoiadb\driver\java\src\test\java\com\sequoiadb\test\node\NoteTest.java 
Modified: E:\sequoiadb\driver\java\src\test\resources\test.properties 
Sending content: E:\sequoiadb\driver\java\src\test\java\com\sequoiadb\test\node\NoteTest.java 
Sending content: E:\sequoiadb\driver\java\src\test\resources\test.properties 
Sending content: E:\sequoiadb\driver\java\src\main\java\com\sequoiadb\base\Sequoiadb.java 
Sending content: E:\sequoiadb\driver\java\src\test\java\com\sequoiadb\test\bug\Bug_JIRA_.java 
Sending content: E:\sequoiadb\driver\java\src\main\java\com\sequoiadb\base\DBCollection.java 
Sending content: E:\sequoiadb\driver\java\src\test\java\com\sequoiadb\test\common\Constants.java 
Sending content: E:\sequoiadb\driver\java\src\test\java\com\sequoiadb\test\TestConfig.java 
```
```
v2.6 
Command: Commit 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\CollectionSpace.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\DBCollection.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\DBCursor.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\DBDataCenter.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\DBLob.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\Node.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\ReplicaGroup.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\ReplicaNode.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\SDBMessage.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\Sequoiadb.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\SequoiadbConstants.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\SequoiadbDatasource.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\SequoiadbOption.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\net\ConfigOptions.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\net\ConnectionTCPImpl.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\net\IConnection.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\net\ServerAddress.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\util\ByteOutputBuffer.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\util\ByteOutputBuffer1.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\util\Helper.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\util\SDBMessageHelper.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\util\logger.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\resources\errors.properties 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\Sequoiadb.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\resources\errors.properties 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\SDBMessage.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\DBDataCenter.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\util\SDBMessageHelper.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\util\ByteOutputBuffer.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\DBLob.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\util\Helper.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\util\logger.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\DBCursor.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\net\IConnection.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\ReplicaNode.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\Node.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\SequoiadbDatasource.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\net\ServerAddress.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\ReplicaGroup.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\net\ConfigOptions.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\DBCollection.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\util\ByteOutputBuffer1.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\net\ConnectionTCPImpl.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\SequoiadbOption.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\CollectionSpace.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\SequoiadbConstants.java 
```

【check list】 
检查项结论（NA/N/Y）
修改代码是否全部提交SVN主干和分支版本且符合代码规范--- v2.9，v2.8，v2.6 
是否用例全部通过-------------------------------------- Y 
是否清淅标明重现步骤---------------------------------- NA 
修改代码是否通过检视---------------------------------- N 
资料是否变更------------------------------------------ NA 
接口是否变更------------------------------------------ N 
相关驱动是否变更或添加-------------------------------- NA 
是否保持向前兼容-------------------------------------- Y 
是否影响安装部署和监控-------------------------------- NA 
是否对影响性能---------------------------------------- NA 
sdb shell帮助信息是否更新----------------------------- NA

Baberir commented:
【检视结果】ok

FlorentinoJH commented:
请安排测试

MarekZZY commented:
请验证

TorstenSQ commented:
【验证结果】 
问题已解决 
【验证步骤】 
1、创建域，名为domain； 
2、在sdb shell执行`db.snapshot( SDB_SNAP_CONTEXTS, { Role: "coord" } )`，并记录上下文数量； 
3、调试Java程序如下： 
 ```
   @Test 
    public void test() { 
        Sequoiadb db = new Sequoiadb("192.168.20.67:11810", "", ""); 
        System.out.println(db.isDomainExist("domain")); 
        db.disconnect(); 
    } 
```
    在isDomainExist处打断点； 
4、到达断点时，重复步骤2； 
5、检查结果，步骤2与步骤4记录的上下文数量相同； 
【测试用例】 
无 
【备注】 
1、回归时引擎必须用2.9的，因为之前的快照无法查看coord上的上下文； 
2、在问题版本中，步骤4的上下文数，会比步骤2中的多一个； 
3、如果只是覆盖isDomainExist()一个方法中的内部游标，感觉没必要，如果要覆盖驱动中所有有游标的地方，范围又太广了。因此没有添加用例；

TorstenSQ commented:
4、除isDomainExist()方法外，其它地方的游标使用的是while(cursor.hasNext()) {...}，这种用法虽然没有主动关闭游标，但是在hasNext()为false的时候，上下文会自动关闭。因此正常情况无法检测是否有泄露，因此不测试；

TorstenSQ commented:
【补充回归版本】 
```
sequoiadb-driver-2.9-r34284.jar 
sequoiadb-driver-2.6.5-r34261.jar
```

TorstenSQ commented:
请检视

AydenYYD commented:
象这种点抛异常，会不会有问题呢？

AydenYYD commented:
见附图，基于取完会关游标，如果是抛异常的情况呢？

TorstenSQ commented:
如附件，在部分地方(如DBCollection.splitAsync())调用 hasNext() 没有放到 try-finally 中。在异常情况下，hasNext() 抛异常可能导致游标泄漏。

TorstenSQ commented:
回归不通过。修改不完全，请开发补充修改。

Bacchusisi commented:
针对回归提出的问题，重新修改。 

【修改】 
所有对游标的操作都放在try{ }finally{ }中进行。 

【提交】 
```
v2.9 
Command: Commit 
Modified: E:\sequoiadb\driver\java\src\main\java\com\sequoiadb\base\DBCollection.java 
Modified: E:\sequoiadb\driver\java\src\main\java\com\sequoiadb\base\Sequoiadb.java 
Modified: E:\sequoiadb\driver\java\src\main\java\com\sequoiadb\datasource\SequoiadbDatasource.java 
Modified: E:\sequoiadb\driver\java\src\test\java\com\sequoiadb\test\bug\Bug_JIRA_.java 
Sending content: E:\sequoiadb\driver\java\src\main\java\com\sequoiadb\datasource\SequoiadbDatasource.java 
Sending content: E:\sequoiadb\driver\java\src\main\java\com\sequoiadb\base\Sequoiadb.java 
Sending content: E:\sequoiadb\driver\java\src\test\java\com\sequoiadb\test\bug\Bug_JIRA_.java 
Sending content: E:\sequoiadb\driver\java\src\main\java\com\sequoiadb\base\DBCollection.java 
```


```
v2.8 
Command: Commit 
Modified: F:\sequoiadb_2.8\driver\java\src\main\java\com\sequoiadb\base\DBCollection.java 
Modified: F:\sequoiadb_2.8\driver\java\src\main\java\com\sequoiadb\base\DBDataCenter.java 
Modified: F:\sequoiadb_2.8\driver\java\src\main\java\com\sequoiadb\base\Sequoiadb.java 
Sending content: F:\sequoiadb_2.8\driver\java\src\main\java\com\sequoiadb\base\DBDataCenter.java 
Sending content: F:\sequoiadb_2.8\driver\java\src\main\java\com\sequoiadb\base\DBCollection.java 
Sending content: F:\sequoiadb_2.8\driver\java\src\main\java\com\sequoiadb\base\Sequoiadb.java 
```


```
v2.6 
Command: Commit 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\DBCollection.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\DBDataCenter.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\Sequoiadb.java 
Modified: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\datasource\SequoiadbDatasourceImpl.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\datasource\SequoiadbDatasourceImpl.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\Sequoiadb.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\DBDataCenter.java 
Sending content: E:\sequoiadb2.6\driver\java\src\main\java\com\sequoiadb\base\DBCollection.java 

```


【check list】 
检查项结论（NA/N/Y）
修改代码是否全部提交SVN主干和分支版本且符合代码规范--- v2.9，v2.8，v2.6 
是否用例全部通过-------------------------------------- Y 
是否清淅标明重现步骤---------------------------------- NA 
修改代码是否通过检视---------------------------------- N 
资料是否变更------------------------------------------ Y 
接口是否变更------------------------------------------ Y 
相关驱动是否变更或添加-------------------------------- NA 
是否保持向前兼容-------------------------------------- Y 
是否影响安装部署和监控-------------------------------- NA 
是否对影响性能---------------------------------------- NA 
sdb shell帮助信息是否更新----------------------------- NA

EmilianLLJH commented:
还有问题

FlorentinoJH commented:
请重新处理

Bacchusisi commented:
针对测试提出的问题，已在上述最后一次回复中提交了修改。

FlorentinoJH commented:
已经完成处理， 请重新测试（基于最新的提交）

EmilianLLJH commented:
请验证

TorstenSQ commented:
【验证结果】 
问题已解决 
【回归版本】 
v2.6 r34689 
v2.8 r34688 
v2.9 r34687 
【验证步骤】 
1、使用ide查看DBCursor.hasNext()和DBCursor.getNext()被调用的地方 
2、检查其是否在try-finally范围内 
【测试用例】 
无

TorstenSQ commented:
请检视

AydenYYD commented:
已经检视，象这种是不是需要mock test?
